File: 0210-course-schedule-ii/0210-course-schedule-ii.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function (numCourses, prerequisites) {
    const adjList = new Map();
    const inDegree = new Map();

    for (let i = 0; i < numCourses; i++) {
        adjList.set(i, []);
        inDegree.set(i, 0);
    }

    for (let [to, from] of prerequisites) {
        const list = adjList.get(from);
        list.push(to);
        adjList.set(from, list);
        inDegree.set(to, inDegree.get(to) + 1);
    }

    const q = [];
    for (let [course, degree] of inDegree) {
        if (degree === 0) {
            q.push(course);
        }
    }

    let count = 0;
    const result = [];
    while (q.length > 0) {
        const node = q.shift();
        result.push(node);
        count++;

        for (const neighbor of adjList.get(node)) {
            inDegree.set(neighbor, inDegree.get(neighbor) - 1);

            if (inDegree.get(neighbor) === 0) q.push(neighbor);

        }
    }

    return count === numCourses ? result : [];
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
from collections import deque


class Solution:
    def findOrder(self, numCourses: int, prerequisites: list[list[int]]) -> list[int]:
        adj_list = {}
        in_degree = {}
        q = deque()
        count = 0
        result = []

        for i in range(numCourses):
            in_degree[i] = 0
            adj_list[i] = []

        for to, frm in prerequisites:
            lst = adj_list.get(frm)
            lst.append(to)
            adj_list[frm] = lst
            in_degree[to] = in_degree.get(to) + 1

        for course, degree in in_degree.items():
            if degree == 0:
                q.append(course)

        while q:
            node = q.popleft()
            result.append(node)
            count += 1

            for neighbor in adj_list.get(node):
                in_degree[neighbor] = in_degree.get(neighbor) - 1

                if in_degree.get(neighbor) == 0:
                    q.append(neighbor)

        return result if count == numCourses else []
